### PR TITLE
[DF] Always check that arg types in jitted code are valid

### PR DESCRIPTION
**Pull request by eguiraud**
Snapshot, Cache, Define and Filter did not check that the types of
columns used as arguments in jitted code were valid (in particular,
that the type of custom columns was correctly understood by cling).
This patch provides a function that performs the validity check,
GetValidatedArgTypes, that is used everywhere instead of GetColumnTypes
(now unused and therefore deleted) and ColumnName2ColumnTypeName.

An exception with a user-friendly error message is thrown in case of
error, which fixes ROOT-10458.

### CODE REVIEW

File: tree/dataframe/src/RDFInterfaceUtils.cxx [hunk 1]
```diff
@@ -753,6 +726,29 @@ ColumnNames_t GetValidatedColumnNames(RLoopManager &lm, const unsigned int nColu
...
+                                              TTree *tree, RDataSource *ds, const std::string &context,
+                                              bool vector2rvec)
+{
+   auto toCheckedArgType = [&](const std::string &c) {
+      RDFDetail::RCustomColumnBase *customCol =
+         customColumns.HasName(c) ? customColumns.GetColumns().at(c).get() : nullptr;
+      const auto colType = ColumnName2ColumnTypeName(c, tree, ds, customCol, vector2rvec);
+      if (colType.rfind("CLING_UNKNOWN_TYPE", 0) == 0) { // the interpreter does not know this type
+         const auto msg =
+            "The type of custom column \"" + c + "\" (" + colType.substr(19) +
```
Comment (hageboeck) [inline tree/dataframe/src/RDFInterfaceUtils.cxx:739]: `substr(19)` puzzles me. Why is this a fixed number?

Comment (eguiraud) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:739]: it's `std::string("CLING_UNKNOWN_TYPE_").size()`. what would you suggest?

Comment (hageboeck) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:739]: Oh, now I see it. Ok, it's safe here.

Comment (hageboeck) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:739]: ... unless the string changes at some point. If that might happen, you could do this:
```
      const auto unknownType = "CLING_UNKNOWN_TYPE";
      if (rfind(unknownType)...
            + unknownType
```
It's unlikely, though, that cling type names change, I assume.

Comment (eguiraud) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:739]: `CLING_UNKNOWN_TYPE` is set by RDF itself, so it won't change unless RDF decides it should change. nevertheless I don't like the magic string floating around -- but I don't know of a nicer solution that would not complicate things further. I hope the unit test I added is enough to catch things going out of sync

Comment (eguiraud) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:739]: ok _now_ the test will definitely notice if the diagnostic changes :D

**Pull request merged**